From 7cb266bf6a9a42cf8654b29728d7eb297af05e5b Mon Sep 17 00:00:00 2001
From: Fabian Keil <fk@fabiankeil.de>
Date: Sat, 5 Aug 2017 13:52:58 +0200
Subject: [PATCH 1/2] salt/grains/core.py: Let _hw_data() validate the kenv
 output on FreeBSD-based systems

Previously systems running in bhyve without smbios had a couple
of grains with the value 'kenv: unable to get ...'.

Now the kenv output is only used if it looks legit.
---
 salt/grains/core.py | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git salt/grains/core.py salt/grains/core.py
index 9b4235da10..d1fc2e85bf 100644
--- salt/grains/core.py
+++ salt/grains/core.py
@@ -2072,7 +2072,12 @@ def _hw_data(osdata):
             }
             for key, val in six.iteritems(fbsd_hwdata):
                 value = __salt__['cmd.run']('{0} {1}'.format(kenv, val))
-                grains[key] = _clean_value(key, value)
+                value = _clean_value(key, value)
+                # XXX: Maybe this should be done in _clean_value()
+                if value is not None and value.startswith('kenv: unable to get'):
+                    grains[key] = None
+                else:
+                    grains[key] = value
     elif osdata['kernel'] == 'OpenBSD':
         sysctl = salt.utils.which('sysctl')
         hwdata = {'biosversion': 'hw.version',
-- 
2.14.1


From fc510dad8bd355d43a4f00166f42a73c019793a5 Mon Sep 17 00:00:00 2001
From: Fabian Keil <fk@fabiankeil.de>
Date: Sat, 5 Aug 2017 14:16:07 +0200
Subject: [PATCH 2/2] salt/grains/core.py: Let _hw_data() fallback to the
 hostuuid sysctl ...

... when the FreeBSD-based system doesn't have a hardware uuid.

For systems running in bhyve there may not be a "hardware" uuid
but the hostuuid rc.d script whil generate a hostuuid from other
system properties and populate kern.hostuuid with it.
---
 salt/grains/core.py | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git salt/grains/core.py salt/grains/core.py
index d1fc2e85bf..c6b18ea47e 100644
--- salt/grains/core.py
+++ salt/grains/core.py
@@ -2078,6 +2078,16 @@ def _hw_data(osdata):
                     grains[key] = None
                 else:
                     grains[key] = value
+        if "uuid" not in grains:
+            # If getting the hardware uuid through kenv failed
+            # (for example because the system is running in bhyve
+            # and doesn't have a hardware uuid), use the uuid
+            # generated by the hostid rc.d script.
+            sysctl = salt.utils.which('sysctl')
+            if sysctl:
+                value = __salt__['cmd.run']('{0} -n kern.hostuuid'.format(sysctl))
+                if not value.startswith('sysctl:'):
+                    grains["uuid"] = value
     elif osdata['kernel'] == 'OpenBSD':
         sysctl = salt.utils.which('sysctl')
         hwdata = {'biosversion': 'hw.version',
-- 
2.14.1

